### ISSUE TITLE
"exec" auth client plugin checks whether STDOUT is interactive to decide whether to forward STDIN

### ISSUE THREAD

**Issue by blowfishpro** [kubernetes/kubernetes#98451]
**What happened**:

I use `aws-vault` with AWS EKS to authenticate access to clusters.  You don't need to worry about how `aws-vault` works, the key point is that it prompts the user to enter a 2FA code before running the AWS command that returns a token.

When a `kubectl` command is part of a larger shell script, it fails to promp for a 2FA code, and thus fails to authenticate with the cluster.  I see error messages like this:

```
Enter token for arn:aws:iam::012345678901:mfa/user.email@example.com: aws-vault: error: exec: Failed to get credentials for production: EOF
```

For reference, here is what the user looks like in my kube config:

```yaml
- name: arn:aws:eks:us-west-2:012345678901:cluster/production-0
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - exec
      - production
      - --
      - aws
      - --region
      - us-west-2
      - eks
      - get-token
      - --cluster-name
      - production-0
      command: aws-vault
```

**What you expected to happen**:

`kubectl` should forward STDIN to the shell so that it can prompt for a 2FA code.

**How to reproduce it (as minimally and precisely as possible)**:

* Have a kube config user that uses exec, the command should prompt the user for input before returning credentials
  * It would probably be sufficient to have a script that prompts the user and then returns static credentials
* Run a `kubectl` command within a shell but send STDOUT somewhere else
  * `echo $(kubectl version)` for instance should work

**Anything else we need to know?**:

I think I see what's going on.  The exec plugin checks whether STDOUT is interactive:

https://github.com/kubernetes/kubernetes/blob/e9dba7a627520f89778b367fc0d955776f220638/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L200

but then uses that result to decide whether to forward STDIN:

https://github.com/kubernetes/kubernetes/blob/e9dba7a627520f89778b367fc0d955776f220638/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L399-L401

In the case of running `kubectl` as part of a a larger script (where the result is either ignored or e.g. piped to another command), STDIN is still interactive, but STDOUT is not.

I didn't find anywhere else where this `interactive` variable is used, although it's possible I missed something.  I also can't think of a case where STDIN would be interactive but you wouldn't want to forward it.

Worth noting, based on that code, that STDERR is forwarded regardless.

**Environment**:
- Kubernetes version (use `kubectl version`):
  ```
  Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
  ```
- Cloud provider or hardware configuration: Amazon EKS 1.18
- OS: Mac OS 11.1
- Kernel: `Darwin 20.2.0 Darwin Kernel Version 20.2.0: Wed Dec  2 20:39:59 PST 2020; root:xnu-7195.60.75~1/RELEASE_X86_64 x86_64`
- Install tools: my `kubectl` installation is provided by Docker Desktop if that matters

**Comment by ankeesler** [comment]
I made a reproducer script[0] for what you are experiencing. When I change that line[1] that @blowfishpro mentioned to check if `os.Stdin.Fd()` is a terminal, that reproducer script passes.

I don't have much historical context on why we were checking `os.Stdout` instead of `os.Stdin`. The `git blame` didn't help either. The `v1alpha1` `ExecCredential` API had an `Interactive` field, but that went away in `v1beta1`.

What about the case where `os.Stdin` is a pipe (or any other "raw" way of hooking up standard input to `kubectl`) (i.e., `echo "$MY_TOKEN" | kubectl --user=interactive-exec-plugin get pods`)? `terminal.IsTerminal(int(os.Stdin.Fd()))` would return `false` for that, but `terminal.IsTerminal(int(os.Stdout.Fd()))` would return `true`. Do we want that case to work?

Could we check if _either_ `os.Stdout` or `os.Stdin` are terminals in order to cover this issue's use case while supporting the pipe use case? More generally, should we set `interactive: canReadFrom(os.Stdin)` (not sure how we would do this?). 

Aside: we actually are hoping to add some integration or cmd tests that validate that this stdin-hooking-up works properly, so this is a great preview of what the test should do.

[0] https://gist.github.com/ankeesler/4d40aee765bc84cf2c959608397353e5
[1] https://github.com/kubernetes/kubernetes/blob/e9dba7a627520f89778b367fc0d955776f220638/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L200

**Comment by ankeesler** [comment]
Wonder if @enj has any thoughts.

**Comment by ankeesler** [comment]
Something occurred to me on this...right now, it kinda seems like a mystery to a running exec plugin whether stdin is hooked up correctly. I feel like an exec plugin would like to be able to declare "yes, I use stdin, please forward stdin to me, regardless of whether I am running in a terminal or not" in the kubeconfig. Then, this whole stdin forwarding would be a lot less magical. This makes me wonder why the `ExecCredential.Spec.Interactive` field was removed from `v1alpha1`... @enj do you have any history on why that was? Was there a security worry about this?

**Comment by liggitt** [comment]
in general, I don't see how an exec plugin that consumes stdin can work when combined with uses of kubectl like this:

```
... thing that produces output ... | kubectl apply -f -
```

Making plugins that consume stdin declare that they do, and erroring early if something else in the kubectl invocation also indicates it wants to consume stdin seems like a good idea to me.

**Comment by ankeesler** [comment]
agreed with the above

short term proposal:
- set `interactive` to `terminal.IsTerminal(os.Stdout.Fd()) || terminal.IsTerminal(os.Stdin.Fd())`, so this will fix @blowfishpro's issue, and should provide a proper superset of functionality to exec plugins upon upgrade

long term proposal (if this is appetizing, we can move this to the KEP):
- add `clientcmd.ExecConfig.Interactive *bool` and `clientauthentication.ExecCredential.Spec.Interactive *bool`
- default this field to `terminal.IsTerminal(os.Stdout.Fd())`
- only forward `os.Stdin` to the exec plugin when `clientauthentication.ExecCredential.Spec.Interactive` is `true`

pro: @blowfishpro will be able to force `os.Stdin` to be forwarded his exec plugin reliably
pro: backwards compatible with previous users of `v1beta1`
pro: plugins can declare when they want to use stdin, so this forwarding is more intentional and observable

con: @blowfishpro may have to update his kubeconfig with this new `interactive` field, which may or may not be hard
con: we still have this weird `terminal.IsTerminal(os.Stdout.Fd())` logic (can we remove this in `v1`?)
con: we may be flapping back to an ExecCredential.Spec.Interactive design that was previously not good?
con: stdin can still be stolen from kubectl (@liggit's last point above)

i started to try to fix that last con, but i couldn't get there...closest i got was this, but i'm not a fan:
- when `kubectl` knows it will use `IOStreams.In`[0], assert that it is not using an exec plugin that wants to use interactive...meh

[0] https://github.com/kubernetes/cli-runtime/blob/984374fbd3bd126ba238900a48adb95b9c380f44/pkg/genericclioptions/io_options.go#L29

**Comment by blowfishpro** [comment]
Right now I'm making manual changes to my kubeconfig anyway, so having to add another field doesn't sound too bad.

The case of piping something to kubectl is an interesting one, and I can see how it's ambiguous in a difficult-to-resolve way.

**Comment by ankeesler** [comment]
That's helpful feedback @blowfishpro, thanks.

Another question - do you have a suitable workaround here? The reason why I ask is because I _think_ we would need to do a proper review on the long term proposal above, which would mean it probably wouldn't land until 1.22 (since 1.21 KEP freeze is in 4 days), so if we don't implement the short-term proposal (which seems a little hacky to me?), then this issue wouldn't be fixed until 1.22.

**Comment by blowfishpro** [comment]
I have an okay workaround for now.  The session with MFA is cached so I just need to remember to run some `aws-vault` command directly before running one of these scripts.  It's annoying but I can live with it for a while.

**Comment by ankeesler** [comment]
Alright here is where I am at with desired behavior (forgetting for a second about backwards compatibility):
1. The `exec.Authenticator` should only try to hook up stdin to the exec plugin if the exec plugin declares that it wants to use it
2. The `exec.Authenticator` should return an error if the exec plugin declares that it wants to use stdin and stdin is not a terminal
3. When an exec plugin that wants to read from stdin (but did not declare it) tries to read from stdin, we should try to make sure they get an error message that is helpful (e.g., "stdin was not hooked up to this process, please declare it" - not sure we can communicate this error across the process boundary between the client-go process and the exec plugin but i'm sure we can figure something out)

Just running through some sharp edges in my head:
- Long running client-go daemon process that uses an exec plugin that declares it wants to use stdin should fail quickly with an error saying that the process stdin is not a terminal
- Long running client-go daemon process that uses an exec plugin that does not declare it wants to use stdin but still tries to use stdin should return a helpful error saying that the exec plugin tried to read from stdin and it is not hooked up (again, not sure we can do much here because of the process boundary)
- Short running kubectl process that uses an exec plugin that does declare it wants to use stdin but takes input from pipe (e.g., `... thing that produces output ... | kubectl apply -f -`) should fail quickly with an error saying that the process stdin is not a terminal
- Short running kubectl process that uses an exec plugin that does not declare it wants to use stdin and then tries to use stdin should fail with a helpful error saying that the exec plugin tried to read from stdin and it is not hooked up (again, not sure we can do much here because of the process boundary)
- Short running kubectl process that uses an exec plugin that does declare it wants to use stdin while kubectl is also using it (e.g., `kubectl apply -f -`) should hopefully succeed since the user will input their stuff on stdin to the exec plugin and _then_ input their stuff on stdin to the `-f -` (this one is nasty)
- Short running kubectl process that uses an exec plugin that does declare it wants to use stdin while kubectl stdout is being directed somewhere else (e.g., `echo "$(kubectl version)"`) should hopefully succeeds since the exec plugin should be writing stuff that it wants the user to see to stderr (e.g., `enter password: `) (this issue)
- Short running kubectl process that uses an exec plugin that does declare it wants to use stdin while kubectl stderr is being directed somewhere else (e.g., `kubectl version 2>/dev/null`) is kinda screwed because the user will never see any prompts (e.g., `enter password`)... :(

**Comment by liggitt** [comment]
Thanks for thinking through all of that (that reads like a comment that took about three days to write)!

Being explicit (in a new client exec config version) seems like a good idea.

> 1. The `exec.Authenticator` should only try to hook up stdin to the exec plugin if the exec plugin declares that it wants to use it

I think I agree with this.

> 2. The `exec.Authenticator` should return an error if the exec plugin declares that it wants to use stdin and stdin is not a terminal

Can we reliably detect stdin is not a terminal? Should client-go be the one to error or should it let the plugin error when reading from stdin fails? I could envision plugins requesting stdin for interactive mode but having a fallback approach if unavailable.

> 3. When an exec plugin that wants to read from stdin (but did not declare it) tries to read from stdin, we should try to make sure they get an error message that is helpful (e.g., "stdin was not hooked up to this process, please declare it" - not sure we can communicate this error across the process boundary between the client-go process and the exec plugin but i'm sure we can figure something out)

If we can reliably detect this, and not send misleading info to the plugin (e.g. hook up a dummy stdin that just returns an EOF error), that maybe seems ok.

> * Short running kubectl process that uses an exec plugin that does declare it wants to use stdin but takes input from pipe (e.g., `... thing that produces output ... | kubectl apply -f -`) should fail quickly with an error saying that the process stdin is not a terminal

I think the error is less that stdin is not a terminal and more that two things are trying to consume it.

**Comment by ankeesler** [comment]
Appreciate that comment @liggitt, thanks for helping move this issue along.

> Can we reliably detect stdin is not a terminal?

I am really hoping that we can be confident that `terminal.IsTerminal(int(os.Stdin.Fd()))` returns true if and only if stdin is a terminal. I don't know how we would otherwise detect that stdin is a terminal. I'm certainly not a subject matter expert here, just speaking from experience and trusting the `terminal` package.

>  I could envision plugins requesting stdin for interactive mode but having a fallback approach if unavailable.

Good call. Agreed. I think this is captures in @enj's proposal below.

> hook up a dummy stdin that just returns an EOF error

I hope that `os.DevNull` will do this for us (see https://pkg.go.dev/os/exec#Cmd `.Stdin`), but I have only tested on macOS.

> I think the error is less that stdin is not a terminal and more that two things are trying to consume it.

Mmm, yeah. The "two things are trying to consume it" bit is still bugging me. I haven't figured out how we detect that. I wonder if someone could also run into this with a kubectl plugin that uses stdin...or if someone built their own command line tool on top of `client-go`?

---

@enj pitched me his enum idea which made sense to me. My understanding of his idea was:
- For beta, add an optional field to the kubeconfig exec config where you can specify whether
  - your exec plugin `Always` wants stdin passed to it and the `exec.Authenticator` should fail if stdin is not a terminal
  - your exec plugin `Never` wants stdin passed to it
  - or your exec plugin wants stdin passed to it if and only if stdin is a terminal, but regardless the exec plugin should always be run
- For GA, make the field required.
- The default behavior for this config option in beta would be what we have today (pass stdin to the exec plugin if stdout is a terminal).
- There was some more conversation in Slack here: https://kubernetes.slack.com/archives/C0EN96KUY/p1613767745112500.

I took a pass at implementing the beta functionality here with the hopes that we could submit it as a "bug fix" for 1.21: https://github.com/kubernetes/kubernetes/pull/99310.

**Comment by ankeesler** [comment]
OK, one final pass here to see if we can make progress on this in 1.21: https://github.com/kubernetes/kubernetes/pull/99654.

@enj posited that we may be able to get away with simply changing the line that @blowfishpro originally pointed out. We would still apply a better fix to this in 1.22, but we wanted to hopefully make progress on this in 1.21.

Slack convo here: https://kubernetes.slack.com/archives/C0EN96KUY/p1614713102054200?thread_ts=1614636195.030500&cid=C0EN96KUY.

**Comment by blowfishpro** [comment]
Thank you to everyone who worked/is working on this.  To be honest I didn't expect it to get traction so quickly.  Thanks for working on it, but also for taking the time to consider how other people might interact with it and do it the right way.
